app/src/main/java/org/osservatorionessuno/bugbane/MainActivity.kt

```diff
@@ -30,64 +34,64 @@ import org.osservatorionessuno.bugbane.components.NavigationTabs
-    private var setLacksPermissionsCallback: ((Boolean) -> Unit)? = null
+
+    private val configViewModel : ConfigurationViewModel by lazy {
+        ViewModelFactory.get(application)
+    }
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
-        PRNGFixes.apply()
-
-        // Observers
-        viewModel.watchConnectAdb().observe(this) { isConnected ->
-            if (!isConnected) {
-                setLacksPermissionsCallback?.invoke(true)
-            }
-        }
-
-        viewModel.watchAskPairAdb().observe(this) { resetPairing ->
-            if (resetPairing) {
-                setLacksPermissionsCallback?.invoke(true)
-            }
-        }
-
-        viewModel.watchCommandOutput().observe(this) { output ->
-            // TODO: blibla
-            Toast.makeText(this@MainActivity, output.toString(), Toast.LENGTH_SHORT).show()
-            Log.d("COMMAND OUTPUT", output.toString())
-        }
 
-        // Try auto-connecting
-        viewModel.autoConnect()
+        PRNGFixes.apply()
 
         // Fetch indicators on first launch and schedule daily background updates
         setupIndicatorsUpdates()
 
-        if (!ConfigurationManager.isNotificationPermissionGranted(this) || !ConfigurationManager.isWirelessDebuggingEnabled(
-                this
-            )
-        ) {
-            setLacksPermissionsCallback?.invoke(true)
-        }
-
-        if (!SlideshowManager.hasSeenHomepage(this)) {
-            // On first start, run the SlideshowActivity manually
-            SlideshowActivity.start(this)
-        }
-
         enableEdgeToEdge()
         setContent {
             Theme {
-                MainContent { callback ->
-                    setLacksPermissionsCallback = callback
+                val appState = configViewModel.configurationState.collectAsStateWithLifecycle()
+
+                LaunchedEffect(appState.value) {
+                    if (!SlideshowManager.hasSeenHomepage(applicationContext)) {
+                        // Permissions slideshow
+                        val startPage = AppState.valuesInOrder().indexOf(appState.value)
+                        val intent = Intent(this@MainActivity, SlideshowActivity::class.java)
+                            .putExtra("startPage", startPage)
+                        startActivity(intent)
+                    }
+                    else if (appState.value == AppState.TryAutoConnect) {
+                        Log.d(TAG, "TryAutoConnect")
+                        configViewModel.adbManager.autoConnect()
+                        // Else-if, to avoid conflict with slideshow connection logic
+                    }
+                }
+
+                if (appState.value != AppState.TryAutoConnect) {
+                    MainContent()
+                } else { // Trying autoconnect, which will either lead to success or need repairing
+                    WaitLoadingContent()
```

Comment:
(note: We can deprecate this WaitLoadingContent() part, which basically blocks the UI until the adb autoconnection has either succeeded or failed. The reason it's here is that I wanted to isolate the changes in this refactor (for testing) before adding additional logic (like disabling buttons until autoconnect is completed)